### PR TITLE
Worldpay commit checks update

### PR DESCRIPTION
**Pull request by ao**

### CODE REVIEW

File: lib/active_merchant/billing/gateways/worldpay_online_payments.rb [hunk 1]
```diff
@@ -141,7 +141,23 @@ def commit(method, url, parameters=nil, options = {})
 
           if (raw_response != '')
             response = parse(raw_response)
-            success = !response.key?("httpStatusCode")
+            if type=='token'
+              success = response.key?("token")
+            elsif
```
Comment (ivanfer) [inline lib/active_merchant/billing/gateways/worldpay_online_payments.rb]: prefer `else` here and below

File: lib/active_merchant/billing/gateways/worldpay_online_payments.rb [hunk 2]
```diff
@@ -141,7 +141,23 @@ def commit(method, url, parameters=nil, options = {})
...
+              if response.key?("httpStatusCode")
+                success = false
+              elsif
+                if type=='authorize'
+                  success = response["paymentStatus"]=="AUTHORIZED" ? true: false
```
Comment (ivanfer) [inline lib/active_merchant/billing/gateways/worldpay_online_payments.rb]: don't need the ternary here

File: lib/active_merchant/billing/gateways/worldpay_online_payments.rb [hunk 3]
```diff
@@ -141,7 +141,23 @@ def commit(method, url, parameters=nil, options = {})
...
+                elsif type=='purchase'
+                  success = response["paymentStatus"]=="SUCCESS" ? true: false
+                elsif type=='capture' || type=='refund' || type=='void'
+                  success = true
```
Comment (ivanfer) [inline lib/active_merchant/billing/gateways/worldpay_online_payments.rb]: These types of operations can never fail?

Comment (ao) [inline lib/active_merchant/billing/gateways/worldpay_online_payments.rb]: They don't contain a response body, merely a 200 and skip the httpStatusCode check above

**Comment by aprofeit** [comment]
@ao could you provide a little description of what the intention of these changes are? Do you have any docs to show why these needed to change?

**Comment by ao** [comment]
@aprofeit These changes are being made because the current implementation does not check for full SUCCESS and if a payment fails during security checking/fraud, then it is still considered a SUCCESS.
The paymentStatus needs to be included to make sure it is 100% SUCCESS.

**Comment by ao** [comment]
@aprofeit Any reason this was closed?

**Comment by aprofeit** [comment]
Was merged in https://github.com/activemerchant/active_merchant/commit/86431f5ae539c594fa702431f63d5524f8f15476 :)

**Comment by ivanfer** [comment]
Thanks @ao !

File: test/unit/gateways/worldpay_online_payments_test.rb [hunk 4]
```diff
@@ -22,8 +22,8 @@ def setup
   end
 
   def test_successful_purchase
-    @gateway.expects(:ssl_request).returns(successful_token_response)
     @gateway.expects(:ssl_request).returns(successful_purchase_response)
+    @gateway.expects(:ssl_request).returns(successful_token_response)
```
Comment (duff) [inline test/unit/gateways/worldpay_online_payments_test.rb]: @mrezentes mentioned to me about the order of these lines not seeming to be the same as the order of the implementation, meaning that it looks like the code creates a token first.

I'm also wondering about how `expects` works and whether we need something like this:

https://github.com/activemerchant/active_merchant/blob/master/test/unit/gateways/redsys_test.rb#L168

**Pull request closed without merging**